sycl/source/detail/queue_impl.cpp

```diff
@@ -192,24 +200,20 @@ void queue_impl::wait(const detail::code_location &CodeLoc) {
   TelemetryEvent = instrumentationProlog(CodeLoc, Name, StreamID, IId);
 #endif
 
-  std::vector<std::shared_ptr<event_impl>> Events;
+  vector_class<std::weak_ptr<event_impl>> Events;
+  vector_class<event> USMEvents;
   {
-    std::lock_guard<mutex_class> Guard(MMutex);
-    for (std::weak_ptr<event_impl> &EventImplWeakPtr : MEvents)
-      if (std::shared_ptr<event_impl> EventImplPtr = EventImplWeakPtr.lock())
-        Events.push_back(EventImplPtr);
+    std::lock_guard<mutex_class> Lock(MMutex);
+    Events = std::move(MEvents);
+    USMEvents = std::move(MUSMEvents);
   }
 
-  for (std::shared_ptr<event_impl> &Event : Events) {
-    Event->wait(Event);
-  }
+  for (std::weak_ptr<event_impl> &EventImplWeakPtr : Events)
+    if (std::shared_ptr<event_impl> EventImplPtr = EventImplWeakPtr.lock())
+      EventImplPtr->wait(EventImplPtr);
 
-  for (event &Event : MUSMEvents) {
+  for (event &Event : USMEvents)
     Event.wait();
-  }
-
-  MEvents.clear();
-  MUSMEvents.clear();
```

Comment:
You missed clear() or what reason we shouldn't clear events in the queue_impl::wait() function now?

Reply:
You won't need to clear `MEvent` and `MUSMEvents` after it's moved to another container.
Though, @romanovvlad  may correct me.